### PR TITLE
just: install/remove git hooks

### PR DESCRIPTION
**Pull request by ursuscamp**
Added `just install-git-hooks` and `just remove-git-hooks` (with contrib script). Fix: #2599

### CODE REVIEW

**Review by tcharding — APPROVED** [review]
ACK 7d5e7bc776c0fc0676e742c6d59be58c1f41d996

Thanks for the contribution! If you use `Fix: #2599` instead of `ref` then github will auto close the issue if/when this PR merges.

I wonder if we should call  the commands `git-hooks-install` and `git-hooks-uninstall` so that they are groped together in the output of `just`, which is currently:

```bash
Available recipes:
    build             # Cargo build everything.
    check             # Cargo check everything.
    default
    docsrs *flags     # Generate documentation.
    fmt               # Run cargo fmt
    format            # Check the formatting
    install-git-hooks # Install git hooks
    lint              # Lint everything.
    remove-git-hooks  # Remove githooks
    sane              # Quick and dirty CI useful for pre-push checks.
    update-lock-files # Update the recent and minimal lock files.
```

**Comment by storopoli** [comment]
ACK 7d5e7bc776c0fc0676e742c6d59be58c1f41d996

> If you use Fix: #2599 instead of ref then github will auto close the issue if/when this PR merges.

You don't need to reword your commit, just edit in the PR description.

**Comment by ursuscamp** [comment]
Thank you @tcharding @storopoli

I have implemented your feedback

File: contrib/symlink-githooks.sh
Comment (Kixunil) [inline contrib/symlink-githooks.sh:33]: It'd be nicer to have both as functions.

Also is the hooks directory guaranteed to exist? It might be good idea to run `mkdir -p`/`rmdir --ignore-fail-on-non-empty`.

Comment (tcharding) [reply contrib/symlink-githooks.sh:33]: I did a quick check and on my system `.git/hooks` exists after running `git init`. This whole script is only for newer devs anyways so I wouldn't be too concerned with setups outside the ordinary (for example script does not work with git trees).

Comment (ursuscamp) [reply contrib/symlink-githooks.sh]: I created an `add_symlinks` function to handle the primary case.

File: contrib/symlink-githooks.sh
Comment (storopoli) [inline contrib/symlink-githooks.sh:27]: Shouldn't this also `exit 0` as well?

Comment (ursuscamp) [reply contrib/symlink-githooks.sh:27]: It's the last executed thing so it's implied, but I can add it if we want to be explicit

Comment (storopoli) [reply contrib/symlink-githooks.sh:27]: I think it is best. Because it is more "future-proof" when someone else wants to change this.

Comment (ursuscamp) [reply contrib/symlink-githooks.sh]: It is updated.

**Review by storopoli — APPROVED** [review]
ACK 1c4bf5086112b984ab28e7100930dd03734e87b5

**Review by tcharding — DISMISSED** [review]
ACK 1c4bf5086112b984ab28e7100930dd03734e87b5

File: contrib/symlink-githooks.sh
Comment (Kixunil) [inline contrib/symlink-githooks.sh:22]: I have a serious concern about the security implications. If a maintainers GH account gets compromised a malicious push could update the hooks to do something malicious thus compromising everyone else using them doing `git pull` and trying to make a contribution. I think we should copy them rather than symlink even though it makes updating them more annoying. If we want smooth updates at all costs maybe we should GPG-sign the hooks and instead copy a script that checks a hook and runs it.

Comment (apoelstra) [reply contrib/symlink-githooks.sh:22]: Agreed re "copying vs symlinking". It does make me nervous to have users auto-running stuff from the repo every time they invoke git.

The hooks should maybe hash themselves and compare to the hash in the repo to alert the user if the in-repo hooks have changed, and tell them how to re-do the copy.

Comment (Kixunil) [reply contrib/symlink-githooks.sh:22]: Good idea with hashing, just make sure to say "please review the hook diff" since if people blindly update we're at square one.

Comment (apoelstra) [reply contrib/symlink-githooks.sh:22]: Sure, we should say that. But even if people are blindly updating the situation is way better than if it were automatic. Since there would be a delay giving time for other maintainers to fix the problem and/or raise the alarm.

Comment (ursuscamp) [reply contrib/symlink-githooks.sh:22]: >The hooks should maybe hash themselves and compare to the hash in the repo to alert the user if the in-repo hooks have changed, and tell them how to re-do the copy.

If the worry is that someone would push malicious hooks to the repo, then a symlinked hook would update during the next `git pull` and its hashed version would always match what was in the repo anyway.

Comment (Kixunil) [reply contrib/symlink-githooks.sh:22]: > its hashed version would always match what was in the repo anyway

Not if it's copied.

Comment (ursuscamp) [reply contrib/symlink-githooks.sh:22]: Oh, I misunderstood.

I will change my PR (or leave it) based on whatever everyone decides.

Comment (apoelstra) [reply contrib/symlink-githooks.sh:22]: To clarify: I am in favor of copying instead of linking. So is @Kixunil. We should let @tcharding chime in as well.

Comment (tcharding) [reply contrib/symlink-githooks.sh:22]: Perhaps we should split this work into two parts

1. This PR, and copy don't link
2. Another PR that adds a [post-merge](https://git-scm.com/docs/githooks#_post_merge) hook that compares the local hook (eg `.git/hooks/pre-commit`) with the one in `githooks/` - this is generally useful, turns out my current local hook is stale.

More about (2)

- Will need to be bullet proof, even for different dev set ups (eg don't hard code path `.git` vs `.bare`).
- We can just use `diff`, right? No need for the added complexity of hashing, AFAICT.

Comment (tcharding) [reply contrib/symlink-githooks.sh:22]: (2) probably already exists on the internet, seems like a general problem that someone will have solved already.

Comment (tcharding) [reply contrib/symlink-githooks.sh:22]: > turns out my current local hook is stale.

Quoting myself, this is a bit ironic since I was probably the one that patched the `pre-commit` hook that made my local one stale.

Comment (ursuscamp) [reply contrib/symlink-githooks.sh]: I modified the PR to do a copy instead of linking. Also, I modified the part of the readme that suggested symlinking or setting the githooks config option to the repo's githooks. Lastly, I rebased on top the latest commits.

File: contrib/symlink-githooks.sh
Comment (Kixunil) [inline contrib/symlink-githooks.sh:15]: The functions have stale names.

File: contrib/symlink-githooks.sh
Comment (Kixunil) [inline contrib/symlink-githooks.sh:9]: Stale message.

File: justfile
Comment (Kixunil) [inline justfile:46]: Stale name

File: contrib/symlink-githooks.sh
Comment (ursuscamp) [reply contrib/symlink-githooks.sh]: Yeah good catch

**Review by Kixunil — APPROVED** [review]
ACK 5bb4d28f659458a3e629a3dd8a3db04531680c2c

**Comment by apoelstra** [comment]
Can you use `$(git rev-parse --git-dir)` rather than hardcoding `.git`? This will work no matter what directory the user calls the script from (and it's very easy to do, so I don't mind holding up the PR over it.)

File: README.md
Comment (tcharding) [inline README.md:189]: If you are updating this PR anyway: `s/folder/directory` please

(Petty "we are Linux here not OSX" complaint.)

File: justfile
Comment (tcharding) [inline justfile:44]:
```suggestion
# Install githooks
```

**Comment by tcharding** [comment]
Both my comments are nits, feel free to ignore (and merge).

**Comment by tcharding** [comment]
oh bother, the commands should be `just githooks-install` and `just githooks-remove` also since "githooks" is a single word.

**Comment by tcharding** [comment]
The `git` docs are a bit confusing but, at least on my machine, we want `git rev-parse --git-common-dir` 

```bash
       The following options are modified by --path-format:
 
       --git-dir
           Show $GIT_DIR if defined. Otherwise show the path to the .git directory. The path shown, when relative, is relative to the current working directory.
 
           If $GIT_DIR is not defined and the current directory is not detected to lie in a Git repository or work tree print a message to stderr and exit with nonzero status.
 
       --git-common-dir
           Show $GIT_COMMON_DIR if defined, else $GIT_DIR.
```

**Comment by tcharding** [comment]
So, in my environment, this works:

```bash
#!/usr/bin/env bash

set -euo pipefail

GIT_DIR=$(git rev-parse --git-common-dir)

remove_githooks() {
	for hook in githooks/*
	do
		bn=$(basename "$hook")
		echo "Removing git hook $bn"
		rm "$GIT_DIR/hooks/$bn"
	done
	exit 0
}

add_githooks() {
        local hooks_dir="$GIT_DIR/hooks"

	mkdir -p "$hooks_dir"
	cp -i githooks/* "$hooks_dir"
	exit 0
}

while getopts "r" flag; do
	case $flag in
		r) remove_githooks
		;;
		*) exit 1
		;;
	esac
done

add_githooks # Copy githooks by default (no options provided)
```

Needs testing by someone that uses `.git` and not worktrees.

**Comment by ursuscamp** [comment]
What about

```bash
GIT_DIR=$(git config --get config.hooksPath || git rev-parse --git-common-dir)
```

**Comment by Kixunil** [comment]
> What about this for anyone who has overridden it.

Yes, good thinking!

**Review by Kixunil — APPROVED** [review]
ACK 1b0653314bee605feb3fcb60e04b00bac9c4aa7a

**Review by tcharding — APPROVED** [review]
ACK 1b0653314bee605feb3fcb60e04b00bac9c4aa7a

**Comment by tcharding** [comment]
Nice, thanks man!

**Comment by tcharding** [comment]
FTR CI fail is unrelated to this PR.

**Pull request merged**
